_posts/hacks/2017-04-14-How to sign.md

```diff
@@ -0,0 +1,208 @@
+import org.bouncycastle.cms.CMSProcessableByteArray;
+import org.bouncycastle.cms.CMSSignedData;
+import org.bouncycastle.cms.CMSSignedDataGenerator;
+import org.bouncycastle.cms.CMSTypedData;
+import org.bouncycastle.cms.jcajce.JcaSignerInfoGeneratorBuilder;
+import org.bouncycastle.operator.ContentSigner;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.operator.jcajce.JcaContentSignerBuilder;
+import org.bouncycastle.operator.jcajce.JcaDigestCalculatorProviderBuilder;
+import org.bouncycastle.util.Store;
+
+List certList = new ArrayList();
+CMSTypedData  msg = new CMSProcessableByteArray("Hello world!".getBytes());
+certList.add(signCert);
+Store certs = new JcaCertStore(certList);
+CMSSignedDataGenerator gen = new CMSSignedDataGenerator();
+ContentSigner sha256Signer = new JcaContentSignerBuilder("SHA256withECDSA").build(signKP.getPrivate());
+
+gen.addSignerInfoGenerator(
+                new JcaSignerInfoGeneratorBuilder(
+                     new JcaDigestCalculatorProviderBuilder().build())
+                     .build(sha256Signer, signCert));
+
+      gen.addCertificates(certs);
+
+      CMSSignedData sigData = gen.generate(msg, false);
+```
+
+Note, that you can define, if also the data and not only signature should be in the CMS-Container.
+With this container you got the signature,
+the certificate, that can be used for verifying, digital algorithm, probably the signed data etc.
+It is also possible to create several signatures for the data and put them in the container. It means you can send this kind of signature
+to a third party, for instance, you can sign data, that are provided by a web service and the consumer wants to verify the signature.
+The code to verify a *CMSSignedData* (again JavaDoc of BounceCastle)
+
+```java
+import org.bouncycastle.cert.X509CertificateHolder;
+import org.bouncycastle.cms.CMSException;
+import org.bouncycastle.cms.CMSSignedData;
+import org.bouncycastle.cms.SignerInformation;
+import org.bouncycastle.cms.SignerInformationStore;
+import org.bouncycastle.cms.jcajce.JcaSimpleSignerInfoVerifierBuilder;
+import org.bouncycastle.operator.OperatorCreationException;
+import org.bouncycastle.util.Store;
+
+Store certStore = cmsSignedData.getCertificates();
+SignerInformationStore signers = cmsSignedData.getSignerInfos();
+Collection c = signers.getSigners();
+Iterator it = c.iterator();
+
+while (it.hasNext())
+  {
+    SignerInformation signer = (SignerInformation)it.next();
+    Collection certCollection = certStore.getMatches(signer.getSID());
+    Iterator certIt = certCollection.iterator();
+    X509CertificateHolder cert = (X509CertificateHolder)certIt.next();
+    if (signer.verify(new JcaSimpleSignerInfoVerifierBuilder().build(cert))) {
+           verified++;
+       }
+   }
+```
+
+If you want to use the whole functionality of a JCE implementation you have to install unrestricted policy files for the JVM. If you don't,
```

Comment:
You should explain the basic limitations (key length, strong algorithms) before.